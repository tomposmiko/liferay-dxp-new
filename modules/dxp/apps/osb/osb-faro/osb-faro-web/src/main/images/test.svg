<svg fill="none" height="20" viewBox="0 0 20 20" width="20" xmlns="http://www.w3.org/2000/svg">
	<rect fill="#6B6C7E" height="20" rx="5" width="20" />
	<path d="M14.7653 13.6478C13.308 10.8283 11.8837 9.2493 11.8762 7.75V5.5C11.8762 5.5 12.1619 5.5007 12.2512 5.5C13.2184 5.4918 13.2937 4.0382 12.2512 4H7.75116C6.72389 4.05789 6.77545 5.46133 7.75116 5.5H8.12616V7.75C8.13975 9.25164 6.56006 11.0352 5.23702 13.6478C4.92225 14.2696 5.09452 15.8924 6.99202 16H13.0103C14.6774 15.9386 15.1429 14.3784 14.7653 13.6478ZM12.9285 14.5H7.05577C6.84577 14.5 6.57342 14.2649 6.74077 13.9075L7.37499 12.6252C9.27108 12.6252 9.97022 11.1231 11.5103 11.1241C11.5103 11.1241 13.1483 13.784 13.2058 13.8625C13.4507 14.1979 13.2137 14.5 12.9285 14.5Z" fill="white" />
</svg>