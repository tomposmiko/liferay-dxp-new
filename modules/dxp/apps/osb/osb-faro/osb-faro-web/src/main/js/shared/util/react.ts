import {isEqual} from 'lodash';

export const getDisplayName = WrappedComponent =>
	WrappedComponent.displayName || WrappedComponent.name || 'Component';

type HasChanges = <T>(
	prev: T | object,
	next: T | object,
	...keys: string[]
) => boolean;

/**
 * Compare previous state or props object by provided keys to detect changes.
 */
export const hasChanges: HasChanges = (prev = {}, next = {}, ...keys) => {
	for (const key of keys) {
		if (key in next) {
			const newVal = next[key];

			const prevVal = prev[key];

			if (!isEqual(newVal, prevVal)) {
				return true;
			}
		}
	}

	return false;
};
