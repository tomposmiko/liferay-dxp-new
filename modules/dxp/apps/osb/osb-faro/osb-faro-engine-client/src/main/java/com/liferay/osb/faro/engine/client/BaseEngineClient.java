/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * The contents of this file are subject to the terms of the Liferay Enterprise
 * Subscription License ("License"). You may not use this file except in
 * compliance with the License. You can obtain a copy of the License by
 * contacting Liferay, Inc. See the License for the specific language governing
 * permissions and limitations under the License, including but not limited to
 * distribution rights of the Software.
 *
 *
 *
 */

package com.liferay.osb.faro.engine.client;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

import com.liferay.osb.faro.engine.client.cache.FaroCache;
import com.liferay.osb.faro.engine.client.exception.FaroEngineClientException;
import com.liferay.osb.faro.engine.client.http.client.AuditClientHttpRequestInterceptor;
import com.liferay.osb.faro.engine.client.http.client.AuthenticationClientHttpRequestInterceptor;
import com.liferay.osb.faro.engine.client.http.client.CacheClientHttpRequestInterceptor;
import com.liferay.osb.faro.engine.client.http.client.LoggingClientHttpRequestInterceptor;
import com.liferay.osb.faro.engine.client.http.client.SSLHandshakeExceptionHttpRequestInterceptor;
import com.liferay.osb.faro.engine.client.mixin.ResourceMixin;
import com.liferay.osb.faro.engine.client.model.BulkRequest;
import com.liferay.osb.faro.engine.client.model.Rels;
import com.liferay.osb.faro.engine.client.util.EngineServiceURLUtil;
import com.liferay.osb.faro.engine.client.util.OrderByField;
import com.liferay.osb.faro.engine.client.web.client.ResponseErrorHandler;
import com.liferay.osb.faro.engine.client.web.util.UriTemplateHandler;
import com.liferay.osb.faro.model.FaroProject;
import com.liferay.osb.faro.util.FaroThreadLocal;
import com.liferay.petra.string.CharPool;
import com.liferay.petra.string.StringPool;
import com.liferay.portal.kernel.log.Log;
import com.liferay.portal.kernel.log.LogFactoryUtil;
import com.liferay.portal.kernel.util.ListUtil;
import com.liferay.portal.kernel.util.StringUtil;
import com.liferay.portal.kernel.util.Validator;

import java.net.URI;
import java.net.URISyntaxException;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import org.apache.http.HttpStatus;
import org.apache.http.client.methods.HttpEntityEnclosingRequestBase;
import org.apache.http.client.methods.HttpUriRequest;

import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.cache.Cache;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.hateoas.EntityModel;
import org.springframework.hateoas.Link;
import org.springframework.hateoas.MediaTypes;
import org.springframework.hateoas.mediatype.hal.Jackson2HalModule;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

/**
 * @author Matthew Kong
 * @author Shinn Lok
 */
public abstract class BaseEngineClient {

	public void setEngineURL(String engineURL) {
		_engineURL = engineURL;

		_urlPaths.clear();
	}

	@SuppressWarnings("unchecked")
	protected <T> List<T> bulk(
		FaroProject faroProject, String type, HttpMethod httpMethod,
		List<Object> bodies, TypeReference<T> typeReference,
		List<Map<String, Object>> uriVariablesList) {

		List<T> responses = new ArrayList<>();

		BulkRequest bulkRequest = new BulkRequest();

		List<BulkRequest.Operation> operations = new ArrayList<>();

		String templatedURL = getTemplatedURL(faroProject, type);

		RestTemplate restTemplate = getRestTemplate(faroProject);

		UriTemplateHandler uriTemplateHandler =
			(UriTemplateHandler)restTemplate.getUriTemplateHandler();

		for (int i = 0; i < uriVariablesList.size(); i++) {
			URI uri = uriTemplateHandler.expand(
				templatedURL, uriVariablesList.get(i));

			Object body = null;

			if (ListUtil.isNotEmpty(bodies)) {
				body = bodies.get(i);
			}

			String url = uri.getPath();

			if (Validator.isNotNull(uri.getQuery())) {
				url += StringPool.QUESTION + uri.getQuery();
			}

			operations.add(
				new BulkRequest.Operation(body, httpMethod.name(), url));
		}

		bulkRequest.setOperations(operations);

		ResponseEntity<Object> responseEntity = restTemplate.exchange(
			getTemplatedURL(faroProject, Rels.BULK), HttpMethod.POST,
			new HttpEntity<>(bulkRequest), Object.class,
			getUriVariables(faroProject));

		Map<String, Object> responseBody =
			(Map<String, Object>)responseEntity.getBody();

		if (responseBody != null) {
			for (Object result : (List<Object>)responseBody.get("responses")) {
				try {
					String resultString = objectMapper.writeValueAsString(
						result);

					if (resultString.startsWith(StringPool.OPEN_BRACKET)) {
						responses.add(
							objectMapper.readValue(
								resultString, typeReference));

						continue;
					}

					Map<String, Map<String, Object>> resultMap =
						(Map<String, Map<String, Object>>)result;

					Map<String, Object> embeddedObject = resultMap.get(
						"_embedded");

					if (embeddedObject == null) {
						responses.add(
							objectMapper.readValue(
								resultString, typeReference));

						continue;
					}

					Collection<Object> collection = embeddedObject.values();

					Iterator<Object> iterator = collection.iterator();

					if (!iterator.hasNext()) {
						continue;
					}

					responses.add(
						objectMapper.readValue(
							objectMapper.writeValueAsString(iterator.next()),
							typeReference));
				}
				catch (Exception exception) {
					_log.error(exception, exception);

					responses.add(null);
				}
			}
		}

		return responses;
	}

	@SuppressWarnings("unchecked")
	protected <T> List<T> bulk(
		FaroProject faroProject, String type, HttpMethod httpMethod,
		TypeReference<T> typeReference,
		List<Map<String, Object>> uriVariablesList) {

		return bulk(
			faroProject, type, httpMethod, new ArrayList<>(), typeReference,
			uriVariablesList);
	}

	protected HttpHeaders createHttpHeaders(Map<String, String> headers) {
		HttpHeaders httpHeaders = new HttpHeaders();

		for (Map.Entry<String, String> entry : headers.entrySet()) {
			httpHeaders.set(entry.getKey(), entry.getValue());
		}

		return httpHeaders;
	}

	protected void delete(
			FaroProject faroProject, Map<String, List<String>> queryParameters)
		throws Exception {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		restTemplate.delete(
			getUriString(
				faroProject, "/projects/" + faroProject.getProjectId(),
				queryParameters));
	}

	protected void delete(
			FaroProject faroProject, String type, List<String> ids)
		throws FaroEngineClientException {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		restTemplate.exchange(
			getTemplatedURL(faroProject, type), HttpMethod.DELETE,
			new HttpEntity<>(ids), String.class, getUriVariables(faroProject));
	}

	protected void delete(
			FaroProject faroProject, String type,
			Map<String, Object> uriVariables)
		throws FaroEngineClientException {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		restTemplate.delete(getTemplatedURL(faroProject, type), uriVariables);
	}

	protected void delete(FaroProject faroProject, String type, String id)
		throws FaroEngineClientException {

		if (Validator.isNull(id)) {
			throw new FaroEngineClientException();
		}

		RestTemplate restTemplate = getRestTemplate(faroProject);

		restTemplate.delete(
			getTemplatedURL(faroProject, type),
			getUriVariables(faroProject, id));
	}

	protected <T> T get(
			FaroProject faroProject, Map<String, String> headers, String path,
			Map<String, List<String>> queryParameters, Class<T> responseType,
			Map<String, Object> uriVariables)
		throws Exception {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<T> responseEntity = restTemplate.exchange(
			getUriString(faroProject, path, queryParameters), HttpMethod.GET,
			new HttpEntity(createHttpHeaders(headers)), responseType,
			uriVariables);

		return responseEntity.getBody();
	}

	protected <T> T get(
		FaroProject faroProject, String type,
		ParameterizedTypeReference<T> parameterizedTypeReference,
		Map<String, Object> uriVariables) {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<T> responseEntity = restTemplate.exchange(
			getTemplatedURL(faroProject, type), HttpMethod.GET,
			HttpEntity.EMPTY, parameterizedTypeReference, uriVariables);

		return responseEntity.getBody();
	}

	protected <T> T get(
			FaroProject faroProject, String type, String id,
			Class<T> responseType)
		throws FaroEngineClientException {

		return get(
			faroProject, type, id, responseType,
			getUriVariables(faroProject, id));
	}

	protected <T> T get(
			FaroProject faroProject, String type, String id,
			Class<T> responseType, Map<String, Object> uriVariables)
		throws FaroEngineClientException {

		if (Validator.isNull(id)) {
			throw new FaroEngineClientException();
		}

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<T> responseEntity = restTemplate.getForEntity(
			getTemplatedURL(faroProject, type), responseType, uriVariables);

		return responseEntity.getBody();
	}

	protected Cache getCache() {
		if (!FaroThreadLocal.isCacheEnabled()) {
			return null;
		}

		Cache cache = (Cache)FaroThreadLocal.getCache();

		if (cache != null) {
			return cache;
		}

		cache = new FaroCache();

		FaroThreadLocal.setCache(cache);

		return cache;
	}

	protected String getEngineURL(FaroProject faroProject) {
		if (Validator.isNotNull(_engineURL)) {
			return _engineURL;
		}

		try {
			return EngineServiceURLUtil.getBackendExternalURL(faroProject);
		}
		catch (URISyntaxException uriSyntaxException) {
			_log.error(uriSyntaxException, uriSyntaxException);
		}

		return _OSB_ASAH_BACKEND_URL;
	}

	protected RestTemplate getRestTemplate(FaroProject faroProject) {
		RestTemplateBuilder restTemplateBuilder = new RestTemplateBuilder();

		restTemplateBuilder = restTemplateBuilder.uriTemplateHandler(
			new UriTemplateHandler());

		MappingJackson2HttpMessageConverter
			mappingJackson2HttpMessageConverter =
				new MappingJackson2HttpMessageConverter();

		mappingJackson2HttpMessageConverter.setObjectMapper(objectMapper);

		mappingJackson2HttpMessageConverter.setSupportedMediaTypes(
			Arrays.asList(MediaType.APPLICATION_JSON, MediaTypes.HAL_JSON));

		restTemplateBuilder = restTemplateBuilder.messageConverters(
			mappingJackson2HttpMessageConverter);

		RestTemplate restTemplate = restTemplateBuilder.build();

		restTemplate.setErrorHandler(new ResponseErrorHandler());

		List<ClientHttpRequestInterceptor> clientHttpRequestInterceptors =
			new ArrayList<>();

		clientHttpRequestInterceptors.add(
			new AuditClientHttpRequestInterceptor());
		clientHttpRequestInterceptors.add(
			new AuthenticationClientHttpRequestInterceptor(faroProject));

		Cache cache = getCache();

		if (cache != null) {
			clientHttpRequestInterceptors.add(
				new CacheClientHttpRequestInterceptor(cache));
		}

		clientHttpRequestInterceptors.add(
			new LoggingClientHttpRequestInterceptor());
		clientHttpRequestInterceptors.add(
			new SSLHandshakeExceptionHttpRequestInterceptor());

		restTemplate.setInterceptors(clientHttpRequestInterceptors);

		restTemplate.setRequestFactory(
			new HttpComponentsClientHttpRequestFactory() {

				@Override
				protected HttpUriRequest createHttpUriRequest(
					HttpMethod httpMethod, URI uri) {

					if (httpMethod == HttpMethod.GET) {
						return new HttpEntityEnclosingRequestBase() {
							{
								setURI(uri);
							}

							@Override
							public String getMethod() {
								return HttpMethod.GET.name();
							}

						};
					}

					return super.createHttpUriRequest(httpMethod, uri);
				}

			});

		return restTemplate;
	}

	protected String getTemplatedURL(FaroProject faroProject, String type) {
		String engineURL = getEngineURL(faroProject);

		if (_urlPaths.containsKey(type)) {
			return engineURL + _urlPaths.get(type);
		}

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<EntityModel<?>> responseEntity = restTemplate.exchange(
			engineURL, HttpMethod.GET, null,
			new ParameterizedTypeReference<EntityModel<?>>() {
			},
			getUriVariables(faroProject));

		if (responseEntity.getStatusCodeValue() != HttpStatus.SC_OK) {
			throw new IllegalStateException("Invalid url: " + engineURL);
		}

		EntityModel<?> resource = responseEntity.getBody();

		if (resource == null) {
			return null;
		}

		Optional<Link> link = resource.getLink(type);

		if (!link.isPresent()) {
			throw new IllegalStateException(
				"URL does not exist for type: " + type);
		}

		String href = link.map(
			Link::getHref
		).get();

		_urlPaths.put(
			type,
			StringUtil.removeSubstring(
				href,
				StringUtil.replace(
					engineURL, "{weDeployKey}", faroProject.getWeDeployKey())));

		return href;
	}

	protected String getUriString(
			FaroProject faroProject, String path,
			Map<String, List<String>> queryParameters)
		throws Exception {

		UriComponentsBuilder uriComponentsBuilder =
			UriComponentsBuilder.fromHttpUrl(
				EngineServiceURLUtil.getBackendURL(faroProject, path));

		for (Map.Entry<String, List<String>> entry :
				queryParameters.entrySet()) {

			List<String> value = entry.getValue();

			uriComponentsBuilder.queryParam(
				entry.getKey(), value.toArray(new String[0]));
		}

		return uriComponentsBuilder.toUriString();
	}

	protected Map<String, Object> getUriVariables(FaroProject faroProject) {
		Map<String, Object> uriVariables = new HashMap<>();

		uriVariables.put("weDeployKey", faroProject.getWeDeployKey());

		return uriVariables;
	}

	protected Map<String, Object> getUriVariables(
		FaroProject faroProject, int start, int end) {

		Map<String, Object> uriVariables = getUriVariables(faroProject);

		uriVariables.put("end", end);
		uriVariables.put("start", start);

		return uriVariables;
	}

	protected Map<String, Object> getUriVariables(
		FaroProject faroProject, int cur, int delta,
		List<OrderByField> orderByFields) {

		return getUriVariables(faroProject, cur, delta, orderByFields, null);
	}

	protected Map<String, Object> getUriVariables(
		FaroProject faroProject, int cur, int delta,
		List<OrderByField> orderByFields, String fieldNameContext) {

		Map<String, Object> uriVariables = getUriVariables(faroProject);

		int page = cur - 1;

		if (page < 0) {
			page = 0;
		}

		uriVariables.put("page", page);

		uriVariables.put("size", delta);

		if (orderByFields == null) {
			return uriVariables;
		}

		List<String> sort = new ArrayList<>();

		for (OrderByField orderByField : orderByFields) {
			String fieldName = orderByField.getFieldName();

			if (!orderByField.isSystem() && (fieldNameContext != null)) {
				fieldName = StringUtil.replace(
					fieldNameContext, CharPool.QUESTION, fieldName);
			}

			sort.add(fieldName + StringPool.COMMA + orderByField.getOrderBy());
		}

		uriVariables.put("sort", sort);

		return uriVariables;
	}

	protected Map<String, Object> getUriVariables(
		FaroProject faroProject, String id) {

		Map<String, Object> uriVariables = getUriVariables(faroProject);

		uriVariables.put("id", id);

		return uriVariables;
	}

	protected <T> T patch(
		FaroProject faroProject, String type, Object object,
		Class<T> responseType, Map<String, Object> uriVariables) {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		return restTemplate.patchForObject(
			getTemplatedURL(faroProject, type), object, responseType,
			uriVariables);
	}

	protected <T> T patch(
		FaroProject faroProject, String type, String id, Object object,
		Class<T> responseType) {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		return restTemplate.patchForObject(
			getTemplatedURL(faroProject, type), object, responseType,
			getUriVariables(faroProject, id));
	}

	protected <T> T post(
			FaroProject faroProject, Map<String, String> headers, String path,
			Map<String, List<String>> queryParameters, Object requestBody,
			Class<T> responseType, Map<String, Object> uriVariables)
		throws Exception {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<T> responseEntity = restTemplate.exchange(
			getUriString(faroProject, path, queryParameters), HttpMethod.POST,
			new HttpEntity<>(requestBody, createHttpHeaders(headers)),
			responseType, uriVariables);

		return responseEntity.getBody();
	}

	protected <T> T post(
		FaroProject faroProject, String type, Object object,
		Class<T> responseType) {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<T> responseEntity = restTemplate.postForEntity(
			getTemplatedURL(faroProject, type), object, responseType,
			getUriVariables(faroProject));

		return responseEntity.getBody();
	}

	protected <T> T post(
		FaroProject faroProject, String type, Object object,
		Class<T> responseType, Map<String, Object> uriVariables) {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<T> responseEntity = restTemplate.postForEntity(
			getTemplatedURL(faroProject, type), object, responseType,
			uriVariables);

		return responseEntity.getBody();
	}

	protected <T> T post(
		FaroProject faroProject, String type, Object object,
		ParameterizedTypeReference<T> responseType,
		Map<String, Object> uriVariables) {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<T> responseEntity = restTemplate.exchange(
			getTemplatedURL(faroProject, type), HttpMethod.POST,
			new HttpEntity<>((T)object), responseType, uriVariables);

		return responseEntity.getBody();
	}

	protected <T> T post(
			FaroProject faroProject, String type, String id,
			Class<T> responseType)
		throws FaroEngineClientException {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<T> responseEntity = restTemplate.postForEntity(
			getTemplatedURL(faroProject, type), null, responseType,
			getUriVariables(faroProject, id));

		return responseEntity.getBody();
	}

	protected <T> T put(
		FaroProject faroProject, String type, Object object,
		Class<T> responseType, Map<String, Object> uriVariables) {

		RestTemplate restTemplate = getRestTemplate(faroProject);

		ResponseEntity<T> responseEntity = restTemplate.exchange(
			getTemplatedURL(faroProject, type), HttpMethod.PUT,
			new HttpEntity<>((T)object), responseType, uriVariables);

		return responseEntity.getBody();
	}

	protected static final ObjectMapper objectMapper = new ObjectMapper() {
		{
			addMixIn(EntityModel.class, ResourceMixin.class);
			configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
			registerModule(new Jackson2HalModule());
		}
	};

	private static final String _OSB_ASAH_BACKEND_URL = System.getenv(
		"OSB_ASAH_BACKEND_URL");

	private static final Log _log = LogFactoryUtil.getLog(
		BaseEngineClient.class);

	private static final Map<String, String> _urlPaths = new HashMap<>();

	private String _engineURL;

}