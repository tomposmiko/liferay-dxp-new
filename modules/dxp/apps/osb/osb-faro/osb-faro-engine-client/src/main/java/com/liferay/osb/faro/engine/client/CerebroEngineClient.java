/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * The contents of this file are subject to the terms of the Liferay Enterprise
 * Subscription License ("License"). You may not use this file except in
 * compliance with the License. You can obtain a copy of the License by
 * contacting Liferay, Inc. See the License for the specific language governing
 * permissions and limitations under the License, including but not limited to
 * distribution rights of the Software.
 *
 *
 *
 */

package com.liferay.osb.faro.engine.client;

import com.liferay.osb.faro.model.FaroProject;

import java.util.Date;
import java.util.Optional;

/**
 * @author Marcellus Tavares
 */
public interface CerebroEngineClient {

	public long getPageViews(
			FaroProject faroProject, Optional<Date> fromDateOptional,
			Optional<Date> toDateOptional)
		throws Exception;

	public String getSiteMetrics(
			String channelId, FaroProject faroProject, String interval,
			int rangeKey)
		throws Exception;

	public boolean isCustomEventsLimitReached(FaroProject faroProject)
		throws Exception;

	public void updateTimeZone(FaroProject faroProject) throws Exception;

}