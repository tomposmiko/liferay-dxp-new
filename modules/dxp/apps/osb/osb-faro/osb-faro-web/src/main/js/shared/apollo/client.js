import cache from './cache';
import Uri from 'metal-uri';
import {ApolloClient} from 'apollo-client';
import {
	CommerceAverageOrderValueResolver,
	CommerceAverageRevenuePerAccountResolver,
	CommerceIncompleteOrdersResolver,
	CommerceTotalOrderValueResolver,
	CustomAssetsListResolver,
	EventAnalysisListResolver,
	ExperimentResolver as Experiment
} from './resolvers';
import {get} from 'lodash';
import {HttpLink} from 'apollo-link-http';
import {onError} from 'apollo-link-error';
import {reloadPage} from 'shared/util/router';

const groupIdRegex = /^\/workspace\/([a-z0-9._-]+)/;

/**
 * Fecth With Project ID
 * @param {string} uri
 * @param {object} options
 */
const fetchWithGroupId = (uri, options) => {
	const currentUri = new Uri(window.location.href);
	const {operationName} = JSON.parse(options.body);
	const pathname = currentUri.getPathname();
	const matches = pathname.match(groupIdRegex);

	if (matches !== null && matches.length > 1) {
		const groupId = matches[1];

		return fetch(
			`${uri}?opname=${operationName}&projectGroupId=${groupId}`,
			options
		);
	}

	return fetch(`${uri}/&opname=${operationName}`, options);
};

const client = new ApolloClient({
	addTypename: true,
	cache,
	defaultOptions: {
		watchQuery: {
			notifyOnNetworkStatusChange: true
		}
	},
	link: HttpLink.from([
		onError(({operation}) => {
			const status = get(operation.getContext(), ['response', 'status']);

			if (status === 401) {
				reloadPage();
			}
		}),
		new HttpLink({
			credentials: 'same-origin',
			fetch: fetchWithGroupId,
			uri: '/o/cerebro/graphql'
		})
	]),
	resolvers: {
		Experiment,
		Query: {
			dashboards(_, params) {
				return CustomAssetsListResolver(params);
			},
			eventAnalysisList(_, params) {
				return EventAnalysisListResolver(params);
			},
			orderAccountAverageCurrencyValues(_, params) {
				return CommerceAverageRevenuePerAccountResolver(params);
			},
			orderAverageCurrencyValues(_, params) {
				return CommerceAverageOrderValueResolver(params);
			},
			orderIncompleteCurrencyValues(_, params) {
				return CommerceIncompleteOrdersResolver(params);
			},
			orderTotalCurrencyValues(_, params) {
				return CommerceTotalOrderValueResolver(params);
			}
		}
	}
});

export default client;
