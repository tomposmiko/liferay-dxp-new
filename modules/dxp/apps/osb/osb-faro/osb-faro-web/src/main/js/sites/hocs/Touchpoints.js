import Card from 'shared/components/Card';
import getMetricsMapper from 'shared/hoc/mappers/metrics';
import React from 'react';
import TouchpointsQuery, {
	TOUCHPOINTS_QUERY_TEST
} from 'shared/queries/TouchpointsQuery';
import URLConstants from 'shared/util/url-constants';
import {
	compose,
	withBaseResults,
	withQueryPagination,
	withQueryRangeSelectors
} from 'shared/hoc';
import {createOrderIOMap, VISITORS_METRIC} from 'shared/util/pagination';
import {getRangeSelectorsFromQuery} from 'shared/util/util';
import {graphql} from '@apollo/react-hoc';
import {
	metricsListColumns,
	sitePagesListColumns
} from 'shared/util/table-columns';
import {RangeKeyTimeRanges} from 'shared/util/constants';
import {Routes} from 'shared/util/router';
import {Sizes} from 'shared/util/constants';

// LRAC-6976 POC TEMP
const withData = (useDB = false) => () =>
	graphql(
		useDB ? TOUCHPOINTS_QUERY_TEST : TouchpointsQuery,
		getMetricsMapper(result => ({
			items: result.pages.assetMetrics,
			total: result.pages.total
		}))
	);

// LRAC-6976 POC TEMP
const getTableWithData = useDB => {
	const TableWithData = withBaseResults(withData(useDB), {
		emptyDescription: (
			<>
				<span className='mr-1'>
					{Liferay.Language.get(
						'check-back-later-to-verify-if-data-has-been-received-from-your-data-sources,-or-you-can-try-a-different-date-range'
					)}
				</span>

				<a
					href={URLConstants.SitesDashboardPages}
					key='DOCUMENTATION'
					target='_blank'
				>
					{Liferay.Language.get('learn-more-about-pages')}
				</a>
			</>
		),
		emptyIcon: {
			border: false,
			size: Sizes.XXXLarge,
			symbol: 'ac-satellite'
		},
		emptyTitle: Liferay.Language.get('there-are-no-pages-found'),
		getColumns: ({
			router: {
				params: {channelId, groupId},
				query
			}
		}) => {
			const rangeSelectors = getRangeSelectorsFromQuery(query);

			return [
				sitePagesListColumns.getTitleUrl({
					channelId,
					groupId,
					rangeSelectors,
					route: Routes.SITES_TOUCHPOINTS_OVERVIEW
				}),
				metricsListColumns.visitorsMetric,
				metricsListColumns.viewsMetric,
				metricsListColumns.avgTimeOnPageMetric,
				metricsListColumns.bounceRateMetric,
				metricsListColumns.entrancesMetric,
				metricsListColumns.exitRateMetric
			];
		},
		legacyDropdownRangeKey: false,
		rowIdentifier: ['assetId', 'assetTitle']
	});

	return TableWithData;
};

const TableWithDataNewDB = getTableWithData(true);
const TableWithDataOldDB = getTableWithData();

// TODO: look intot he default range key
const Touchpoints = ({router, ...otherProps}) => {
	const TableWithData =
		router.query.useDB === 'true' ? TableWithDataNewDB : TableWithDataOldDB;

	return (
		<Card className='site-touchpoints-root' pageDisplay>
			<TableWithData
				{...otherProps}
				entityLabel={Liferay.Language.get('pages')}
				router={router}
			/>
		</Card>
	);
};

export default compose(
	withQueryPagination({initialOrderIOMap: createOrderIOMap(VISITORS_METRIC)}),
	withQueryRangeSelectors({rangeKey: RangeKeyTimeRanges.Last30Days})
)(Touchpoints);
