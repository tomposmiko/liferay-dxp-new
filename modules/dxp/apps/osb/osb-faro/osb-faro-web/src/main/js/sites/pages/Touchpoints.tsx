import React from 'react';
import Touchpoints from 'sites/hocs/Touchpoints';

interface ITouchpointsPageProps {
	router: object;
}

export default class TouchpointsPage extends React.Component<ITouchpointsPageProps> {
	render() {
		const {router} = this.props;

		return (
			<div className='sites-dashboard-touchpoints-list-root'>
				<div className='row'>
					<div className='col-xl-12'>
						<Touchpoints router={router} />
					</div>
				</div>
			</div>
		);
	}
}
