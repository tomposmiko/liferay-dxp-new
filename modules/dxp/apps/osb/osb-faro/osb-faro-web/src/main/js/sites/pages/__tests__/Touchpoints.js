import client from 'shared/apollo/client';
import React from 'react';
import Touchpoints from '../Touchpoints';
import {ApolloProvider} from '@apollo/react-components';
import {render} from '@testing-library/react';
import {StaticRouter} from 'react-router';
import {waitForLoading} from 'test/helpers';

jest.unmock('react-dom');

describe('Sites Dashboard Touchpoints Page', () => {
	it('render', async () => {
		const {container} = render(
			<ApolloProvider client={client}>
				<StaticRouter>
					<Touchpoints router={{params: {}, query: {}}} />
				</StaticRouter>
			</ApolloProvider>
		);

		await waitForLoading(container);

		jest.runAllTimers();

		expect(container).toMatchSnapshot();
	});
});
