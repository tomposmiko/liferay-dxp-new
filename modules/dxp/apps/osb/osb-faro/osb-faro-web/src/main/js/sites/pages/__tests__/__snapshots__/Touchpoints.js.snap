// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sites Dashboard Touchpoints Page render 1`] = `
<div>
  <div
    class="sites-dashboard-touchpoints-list-root"
  >
    <div
      class="row"
    >
      <div
        class="col-xl-12"
      >
        <div
          class="card card-root site-touchpoints-root page-display"
        >
          <div
            class="d-flex flex-column flex-grow-1 justify-content-between"
          >
            <div
              class="toolbar-root"
            >
              <nav
                class="navbar navbar-root management-bar management-bar-light navbar-expand"
              >
                <div
                  class="container-fluid"
                >
                  <ul
                    class="nav-root front-nav navbar-nav"
                    role="tablist"
                  />
                  <div
                    class="navbar-form navbar-form-autofit"
                  >
                    <div
                      class="input-group search"
                    >
                      <div
                        class="input-group-item"
                      >
                        <input
                          class="input-root form-control input-group-inset input-group-inset-after"
                          placeholder="Search"
                          type="text"
                          value=""
                        />
                        <div
                          class="input-group-inset-item input-group-inset-item-after"
                        >
                          <button
                            class="button-root btn btn-unstyled"
                            type="button"
                          >
                            <svg
                              class="lexicon-icon lexicon-icon-search icon-root"
                              role="presentation"
                            >
                              <use
                                xlink:href="#search"
                              />
                            </svg>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="no-results-root flex-grow-1"
                  >
                    <div
                      class="no-results-content"
                    >
                      <h4
                        class="no-results-title"
                      >
                        Sorry, an error occurred.
                      </h4>
                    </div>
                  </div>
                </div>
              </nav>
            </div>
            <div
              class="error-display-root flex-grow-1 error-spacer no-results-root flex-grow-1"
            >
              <div
                class="no-results-content"
              >
                <h4
                  class="no-results-title"
                >
                  An unexpected error occurred.
                </h4>
                <button
                  class="button-root btn btn-secondary"
                  type="button"
                >
                  Reload
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
